### PR TITLE
[e2e] Check ClusterOperator Status upon initialization

### PR DESCRIPTION
**Pull request by galletti94**
This test checks each of the ClusterOperator Conditions in its Status and
ensures that these match the expected Conditions when the marketplace operator
is available and healthy.

### CODE REVIEW

File: test/testgroups/clusteroperatortests.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+	"testing"
+
+	"github.com/operator-framework/operator-marketplace/test/testsuites"
+)
+
+// ClusterOperatorTestGroup runs test suites that check the status of the Cluster Operator
+func ClusterOperatorTestGroup(t *testing.T) {
+	// Run the test suites.
+	t.Run("cluster-operator-status-on-startup-test-suite", testsuites.ClusterOperatorStatusOnStartup)
+}
```
Comment (aravindhp) [inline test/testgroups/clusteroperatortests.go]: Please add new line

File: test/testsuites/clusteroperatorstatustests.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	ctx := test.NewTestCtx(t)
+	defer ctx.Cleanup()
+
+	// Get global framework variables.
+	client := test.Global.Client
+
+	// Get namespace
+	namespace, err := test.NewTestCtx(t).GetNamespace()
+	if err != nil {
+		t.Fatalf("Could not get namespace: %v", err)
```
Comment (aravindhp) [inline test/testsuites/clusteroperatorstatustests.go]: Please use `require` or `assert` appropriately as per #177. This comment applies to all places where you are checking for errors.

File: test/testsuites/clusteroperatorstatustests.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+		t.Fatalf("ClusterOperator never reached expected status: %v", err)
+	}
+}
+
+// clusterOperatorCheckTypeStatus checks each condition (type & status) against expectedTypeStatus
+func clusterOperatorCheckTypeStatus(client test.FrameworkClient, clusterOperatorName string, namespace string, expectedTypeStatus map[configv1.ClusterStatusConditionType]configv1.ConditionStatus) error {
+	// Get ClusterOperator resource with clusterOperatorName name
+	namespacedName := types.NamespacedName{Name: clusterOperatorName, Namespace: namespace}
+	result := &configv1.ClusterOperator{}
+	// No need to poll here since the ClusterOperator resource should be there when tests are launched
```
Comment (aravindhp) [inline test/testsuites/clusteroperatorstatustests.go]: Why shouldn't you poll? It is possible the tests run and for some reason the operator is coming up slowly on the cluster.

File: test/testsuites/clusteroperatorstatustests.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	// Get namespace
+	namespace, err := test.NewTestCtx(t).GetNamespace()
+	if err != nil {
+		t.Fatalf("Could not get namespace: %v", err)
+	}
+
+	// Check that the ClusterOperator resource has the correct status
+	clusterOperatorName := "marketplace"
+	expectedTypeStatus := map[configv1.ClusterStatusConditionType]configv1.ConditionStatus{
+		"Progressing": "False",
```
Comment (aravindhp) [inline test/testsuites/clusteroperatorstatustests.go]: Aren't there constants that you can use here from `configv1`?

Comment (galletti94) [reply test/testsuites/clusteroperatorstatustests.go]: Great point!

File: test/testsuites/clusteroperatorstatustests.go (hunk 3 above)
Comment (galletti94) [reply test/testsuites/clusteroperatorstatustests.go]: ok I agree - what timeout length would you advise here?

Comment (aravindhp) [reply test/testsuites/clusteroperatorstatustests.go]: I would say at least 5-10 minutes. But others should chime in here too.

Comment (galletti94) [reply test/testsuites/clusteroperatorstatustests.go]: Thanks! @awgreene @kevinrizza @anik120 what do you think?

File: test/testsuites/clusteroperatorstatustests.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	require.NoError(t, err, "ClusterOperator never reached expected status")
+}
+
+// clusterOperatorCheckTypeStatus checks each condition (type & status) against expectedTypeStatus
+func clusterOperatorCheckTypeStatus(client test.FrameworkClient, clusterOperatorName string, namespace string, expectedTypeStatus map[configv1.ClusterStatusConditionType]configv1.ConditionStatus) error {
+	// Get ClusterOperator resource with clusterOperatorName name
+	namespacedName := types.NamespacedName{Name: clusterOperatorName, Namespace: namespace}
+	result := &configv1.ClusterOperator{}
+
+	// Poll in case the ClusterOperator resource is not immediately present on the cluster
```
Comment (aravindhp) [inline test/testsuites/clusteroperatorstatustests.go]: It is not enough to poll for the presence. It is possible that the `ClusterOperator` is present but not in the state we expect it to be. That is what you should be polling for.

File: test/testsuites/clusteroperatorstatustests.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		return err
+	}
+	// Check that each ClusterOperator ConditionType has a ConditionStatus matching expectedTypeStatus
+	for _, condition := range result.Status.Conditions {
+		if expectedTypeStatus[condition.Type] != condition.Status {
```
Comment (aravindhp) [inline test/testsuites/clusteroperatorstatustests.go]: This should be an `assert`.

File: test/testsuites/clusteroperatorstatustests.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+// clusterOperatorCheckTypeStatus checks each condition (type & status) against expectedTypeStatus
+func clusterOperatorCheckTypeStatus(client test.FrameworkClient, clusterOperatorName string, namespace string, expectedTypeStatus map[configv1.ClusterStatusConditionType]configv1.ConditionStatus) error {
+	// Get ClusterOperator resource with clusterOperatorName name
+	namespacedName := types.NamespacedName{Name: clusterOperatorName, Namespace: namespace}
+	result := &configv1.ClusterOperator{}
+
+	// Poll in case the ClusterOperator resource is not immediately present on the cluster
+	RetryInterval := time.Second * 5
+	Timeout := time.Minute * 5
+	err := wait.Poll(RetryInterval, Timeout, func() (done bool, err error) {
```
Comment (kevinrizza) [inline test/testsuites/clusteroperatorstatustests.go]: Why can't you just use the helper method here?

`helpers.WaitForResult(client, result, clusterOperatorName, namespace)`

File: test/testsuites/clusteroperatorstatustests.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		}
+
+		return true, nil
+	})
+	if err != nil {
```
Comment (aravindhp) [inline test/testsuites/clusteroperatorstatustests.go]: This should be `require`. (I think :-))

File: test/e2e/main_test.go [hunk 9]
```diff
@@ -64,4 +66,16 @@ func initTestingFramework(t *testing.T) {
 	if err != nil {
 		t.Fatalf("failed to add CatalogSource custom resource scheme to framework: %v", err)
 	}
+	// Add (configv1) ClusterOperator to framework scheme
+	clusterOperator := &configv1.ClusterOperator{
+		TypeMeta: metav1.TypeMeta{
+			Kind: "ClusterOperator",
```
[the remark this answers is not in the document]
Comment (galletti94) [reply test/e2e/main_test.go]: I get `ClusterOperator` undefined if I remove the quotes. And `configv1` doesn't seem to have `CluterOperatorKind` defined.

[the remark this answers is not in the document]
Comment (aravindhp) [reply test/e2e/main_test.go]: Duh wrong suggestion on my part. Works only in YAML :-)

File: test/testsuites/clusteroperatorstatustests.go [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+			return false, err
+		}
+		for _, condition := range result.Status.Conditions {
+			if expectedTypeStatus[condition.Type] != condition.Status {
+				return false, fmt.Errorf("Expecting condition type %v of status %v but got %v", condition.Type, expectedTypeStatus[condition.Type], condition.Status)
+			}
+		}
+		return true, nil
+	})
+	require.NoError(t, err, "ClusterOperator never reached expected status")
```
Comment (aravindhp) [inline test/testsuites/clusteroperatorstatustests.go]: This should be an `assert`

**Comment by galletti94** [comment]
@aravindhp @kevinrizza I addressed your comments. Please let me know if it lgty

**Pull request merged**